File: csharp/src/Noie.Nds.Api.Abstractions/Transaction/ConsistencyMode.cs
namespace Noie.Nds.Api.Transaction;

/// <summary>
/// [Index: NDS-CSHARP-CONSISTENCYMODE-000] Consistency mode.
/// 
/// <para>[Semantic] Consistency requirements for applying a transaction.</para>
/// </summary>
public enum ConsistencyMode
{
    /// <summary>
    /// Strong consistency.
    /// Applies immediately; prioritizes correctness.
    /// Suitable for critical economic operations (payments, transfers).
    /// </summary>
    Strong,
    
    /// <summary>
    /// Eventual consistency.
    /// May apply with delay; converges over time.
    /// Suitable for non-critical operations (stats, logs).
    /// </summary>
    Eventual,
    
    /// <summary>
    /// Optimistic consistency.
    /// Based on optimistic concurrency; may fail.
    /// Suitable for high-concurrency scenarios.
    /// </summary>
    Optimistic
}

/// <summary>
/// [Index: NDS-CSHARP-CONSISTENCYMODE-100] ConsistencyMode extensions.
/// </summary>
public static class ConsistencyModeExtensions
{
    /// <summary>
    /// [Index: NDS-CSHARP-CONSISTENCYMODE-110] Parse a consistency mode token.
    /// </summary>
    public static ConsistencyMode FromString(string? str)
    {
        if (string.IsNullOrEmpty(str)) return ConsistencyMode.Strong;
        
        return str.ToUpperInvariant() switch
        {
            "STRONG" => ConsistencyMode.Strong,
            "EVENTUAL" => ConsistencyMode.Eventual,
            "OPTIMISTIC" => ConsistencyMode.Optimistic,
            _ => ConsistencyMode.Strong
        };
    }
    
    /// <summary>
    /// [Index: NDS-CSHARP-CONSISTENCYMODE-120] Convert to proto string token.
    /// </summary>
    public static string ToProtoString(this ConsistencyMode mode)
    {
        return mode switch
        {
            ConsistencyMode.Strong => "STRONG",
            ConsistencyMode.Eventual => "EVENTUAL",
            ConsistencyMode.Optimistic => "OPTIMISTIC",
            _ => "STRONG"
        };
    }
}

File: java/src/main/java/noie/linmimeng/noiedigitalsystem/api/transaction/ConsistencyMode.java
package noie.linmimeng.noiedigitalsystem.api.transaction;

/**
 * [Index] NDS-JAVA-CONSISTENCYMODE-000
 * [Semantic] Consistency requirements for applying a transaction.
 *
 * @since 2.0.0
 */
public enum ConsistencyMode {
    /**
     * Strong consistency.
     * Applies immediately; prioritizes correctness.
     * Suitable for critical economic operations (payments, transfers).
     */
    STRONG,

    /**
     * Eventual consistency.
     * May apply with delay; converges over time.
     * Suitable for non-critical operations (stats, logs).
     */
    EVENTUAL,

    /**
     * Optimistic consistency.
     * Based on optimistic concurrency; may fail.
     * Suitable for high-concurrency scenarios.
     */
    OPTIMISTIC;

    /**
     * @param str string token (case-insensitive)
     * @return parsed mode; defaults to {@link #STRONG} if null or unrecognized
     */
    public static ConsistencyMode fromString(String str) {
        if (str == null) return STRONG;
        try {
            return valueOf(str.toUpperCase());
        } catch (IllegalArgumentException e) {
            return STRONG;
        }
    }
}

